src/AsyncProducers.cpp

```diff
@@ -592,6 +661,98 @@ class TightenProducerConsumerNodes : public IRMutator {
     }
 };
 
+// Update indices to add double buffering.
+class UpdateIndices : public IRMutator {
+    using IRMutator::visit;
+
+    Stmt visit(const Provide *op) override {
+        if (op->name == func_name) {
+            std::vector<Expr> args = op->args;
+            args.push_back(ring_buffer_index);
+            return Provide::make(op->name, op->values, args, op->predicate);
+        }
+        return IRMutator::visit(op);
+    }
+
+    Expr visit(const Call *op) override {
+        if (op->call_type == Call::Halide && op->name == func_name) {
+            std::vector<Expr> args = op->args;
+            args.push_back(ring_buffer_index);
+            return Call::make(op->type, op->name, args, op->call_type, op->func, op->value_index, op->image, op->param);
+        }
+        return IRMutator::visit(op);
+    }
+
+    std::string func_name;
+    Expr ring_buffer_index;
+
+public:
+    UpdateIndices(const string &fn, Expr di)
+        : func_name(fn), ring_buffer_index(std::move(di)) {
+    }
+};
+
+// Inject double buffering.
+class InjectDoubleBuffering : public IRMutator {
+    using IRMutator::visit;
+
+    Stmt visit(const Realize *op) override {
+        Stmt body = mutate(op->body);
+        Function f = env.find(op->name)->second;
+        Region bounds = op->bounds;
+        if (f.schedule().ring_buffer().defined()) {
+            std::string enclosing_loop_var = loop_names.back();
+
+            bounds.emplace_back(0, f.schedule().ring_buffer());
```

Comment:
I am confused about how `bounds` is used here, isn't it a vector<Range>, so 0 is a min and ring_buffer() is an Expr of extent? I guess that makes sense but is a little other than what I would have expected. might be worth adding some comments about the `bounds` semantics for ring buffers if there aren't any already